selftests/unit/test_utils_cloudinit.py

```diff
@@ -0,0 +1,91 @@
+import os
+import shutil
+import tempfile
+import threading
+import unittest     # pylint: disable=C0411
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+
+from six.moves import http_client
+
+from avocado.utils import cloudinit
+from avocado.utils import iso9660
+from avocado.utils import network
+from avocado.utils import data_factory
+
+
+def has_iso_create_write():
+    return iso9660.iso9660(os.devnull, ["create", "write"]) is not None
+
+
+class CloudInit(unittest.TestCase):
+
+    def test_iso_no_create_write(self):
+        with mock.patch('avocado.utils.iso9660.iso9660', return_value=None):
+            self.assertRaises(RuntimeError, cloudinit.iso, os.devnull, "INSTANCE_ID")
+
+
+class CloudInitISO(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp(prefix="avocado_" + __name__)
+
+    @unittest.skipUnless(has_iso_create_write(),
+                         "system lacks support for creating ISO images")
+    def test_iso_no_phone_home(self):
+        path = os.path.join(self.tmpdir, "cloudinit.iso")
+        instance_id = b"INSTANCE_ID"
+        username = b"AVOCADO_USER"
+        password = b"AVOCADO_PASSWORD"
+        cloudinit.iso(path, instance_id, username, password)
+        iso = iso9660.iso9660(path)
+        self.assertIn(instance_id, iso.read("/meta-data"))
+        user_data = iso.read("/user-data")
+        self.assertIn(username, user_data)
+        self.assertIn(password, user_data)
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+
+class PhoneHome(unittest.TestCase):
+
+    def test_phone_home(self):
```

Comment:
There are thre things being done in this test method: 
1) server setup, 
2) test server contacting a URL with bad instance id and 
3) test server with proper instance id. 

If one of these things fail, it will be little bit harder to find out what fails and why. Probably we'll have to open the test code to understand exactly. So, my proposal is to break these things like that:
```python
class PhoneHome(unittest.TestCase):

    def setUpClass(self):
        #  setup the server before all tests and try to reuse it

    def tearDownClass(self):
        #  dispose server

    def test_phone_home_with_wrong_instance_id(self):
        (...)

    def test_phone_home_with_right_instance_id(self):
        (...)
```